Common/Serialize/Serializer.h

```diff
@@ -192,47 +192,63 @@ class CChunkFileReader
 		return (size_t)ptr;
 	}
 
-	// Expects ptr to have at least MeasurePtr bytes at ptr.
+	// If *saved is null, will allocate storage using malloc.
+	// If it's not null, it will be used, but only hope can save you from overruns at the end. For libretro.
 	template<class T>
-	static Error SavePtr(u8 *ptr, T &_class, size_t expected_size)
+	static Error MeasureAndSavePtr(T &_class, u8 **saved, size_t *savedSize)
 	{
-		const u8 *expected_end = ptr + expected_size;
-		PointerWrap p(&ptr, PointerWrap::MODE_WRITE);
+		u8 *ptr = nullptr;
+		PointerWrap p(&ptr, PointerWrap::MODE_MEASURE);
+		_class.DoState(p);
+		_assert_(p.error == PointerWrap::ERROR_NONE);
+
+		size_t measuredSize = p.Offset();
+		u8 *data;
+		if (*saved) {
+			data = *saved;
+		} else {
+			data = (u8 *)malloc(measuredSize);
+			if (!data)
+				return ERROR_BAD_ALLOC;
+		}
+
+		p.RewindForWrite(data);
 		_class.DoState(p);
 
-		if (p.error != PointerWrap::ERROR_FAILURE && (expected_end == ptr || expected_size == 0)) {
+		if (p.CheckAfterWrite()) {
+			*saved = data;
+			*savedSize = measuredSize;
 			return ERROR_NONE;
 		} else {
+			free(data);
 			return ERROR_BROKEN_STATE;
 		}
 	}
 
+	// Duplicate of the above but takes and modifies a vector. Less invasive
+	// than modifying the rewind manager to keep things in something else than vectors.
 	template<class T>
-	static Error MeasureAndSavePtr(T &_class, u8 **saved, size_t *savedSize)
+	static Error MeasureAndSavePtr(T &_class, std::vector<u8> *saved)
 	{
 		u8 *ptr = nullptr;
 		PointerWrap p(&ptr, PointerWrap::MODE_MEASURE);
 		_class.DoState(p);
 		_assert_(p.error == PointerWrap::ERROR_NONE);
 
 		size_t measuredSize = p.Offset();
-		u8 *data = (u8 *)malloc(measuredSize);
-		if (!data)
-			return ERROR_BAD_ALLOC;
-
+		saved->resize(measuredSize);
```

Comment:
Hm, I think this was a problem before but it'd be more likely now (and especially if using zstd.)

If the compress thread is running and didn't finish before this one, the resize might relocate the memory and cause a crash.  Before it would only resize up, which was still a problem, but perhaps less likely.

One solution would be to stall on `compressThread_.join();` before starting a new `Save()`.

Maybe this was the crash you saw before.

-[Unknown]

Reply:
No, the flaky crash I saw before was just the iterator assert, which for whatever reason wasn't triggered every time.

But yes, we should join the compressThread_ before saving (or allow concurrent ones with their own buffers, but kinda overkill). Fixing.